Add video importer

berarma opened the pull request:
Closes godotengine/godot-proposals#7062.

The goal for this PR is being able to import video files from a variety of formats to be used with the built-in VideoStreamPlayer. The imported video files are converted internally to Ogg Theora (OGV) using an external FFmpeg binary so that users don't have to do it manually.

The user has to set  `filesystem/import/video/ffmpeg_path` in the editor settings to the FFmpeg binary in their system.

The accepted video file extensions are:

- mp4
- webm
- mpg
- mpeg
- mkv
- avi
- mov
- wmv

I've chosen these extensions because they're well known. Suggestions welcomed. The supported video codecs will be the ones compatible with these containers and built into the FFmpeg version used.

The import options are:

- Video quality.
- Audio quality.
- Keyframe interval (GOP).
- Scale width.
- Scale height.
- Alpha channel replacement color. (Replaces the alpha channel, if there's one, with a color.)
- Encoding speed.
- Custom FFmpeg arguments.

~~I would like to add the Theora encoding speed level to allow for faster encoding, but it's not an option in FFmpeg. [I've sent a patch to add this option to FFmpeg](https://ffmpeg.org/pipermail/ffmpeg-devel/2025-July/346690.html) but this is my first patch submission... we'll see.~~

## TODO

- [x] Import video files to OGV.
- [x] Imported video files working in the editor.
- [x] Imported video files working in project exports.
- [x] Add speed level option to this importer and FFmpeg.
- [x] Add a setting to disable the video importer when using an external addon for videos?
- [x] Show a warning when using an old FFmpeg version that doesn't support `-speed_level` and importing with a speed setting different to the FFmpeg default.

## Possible improvements?

- Imports can be really slow, as expected, and freeze the editor while importing. Can anything be done about this?
- Create asset library addon or automation to install FFmpeg? Calinou has proposed to create an external tools manager to centralize downloads.

berarma committed:
Add video importer

BlueCube3310 commented on editor/settings/editor_settings.cpp:
```diff
@@ -668,6 +668,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 
 	// Tools (denoise)
 	EDITOR_SETTING_USAGE(Variant::STRING, PROPERTY_HINT_GLOBAL_DIR, "filesystem/tools/oidn/oidn_denoise_path", "", "", PROPERTY_USAGE_DEFAULT)
+	EDITOR_SETTING_USAGE(Variant::STRING, PROPERTY_HINT_GLOBAL_DIR, "filesystem/tools/video/ffmpeg_path", "", "", PROPERTY_USAGE_DEFAULT)
```
I think this should be placed in the "filesystem/import" category, similar to Blender or FBX2GLTF

beicause commented on modules/theora/video_stream_theora.h:
```diff
@@ -155,6 +155,8 @@ class VideoStreamPlaybackTheora : public VideoStreamPlayback {
 
 class VideoStreamTheora : public VideoStream {
 	GDCLASS(VideoStreamTheora, VideoStream);
+	OBJ_SAVE_TYPE(VideoStream); // Saves derived classes with common type so they can be interchanged.
+	RES_BASE_EXTENSION("theoravideostr");
```
Is this needed? It's already saved as `VideoStreamTheora`, isn't it?

beicause commented on modules/theora/resource_importer_video.cpp:
```diff
@@ -0,0 +1,142 @@
+		return FAILED;
+	}
+
+	VideoStreamTheora *stream = memnew(VideoStreamTheora);
+	stream->set_file(p_save_path + ".ogv");
```
This doesn't seem to work. `VideoStreamTheora` can't be saved because it does not have any property to save.
Maybe we should add `file` to its property, and use `r_gen_files->push_back(p_save_path + ".ogv")` to include the `.ogv` file as deps.

berarma replied:
Thank you.

beicause reviewed (commented):
Importer is editor-only, so it should be moved to the `editor` folder, and `SCsub` will include it in editor build.

Also, please change `config.py` to generate the importer docs in `doc_classes` folder.

berarma commented:
> Importer is editor-only, so it should be moved to the `editor` folder, and `SCsub` will include it in editor build.

That was my first thought but I saw the Vorbis importer isn't editor-only and I guessed there was some use for the importers out of the editor. Why?

You're right that having the setting in the editor settings makes more sense, I just lost the scope while fighting with the code.

> Also, please change config.py to generate the importer docs in doc_classes folder.

Sorry, what do I need to change?

lyuma commented:
Just to confirm, because the description doesn't explain what this PR actually does...

From what I understand, this adds importer support for mp4/webm/mpg/mpeg/mkv/avi/mov/wmv/qoi extensions? This is pretty innovative and should make videos much easier to use in Godot

They are transcoded into OGV using some configurable bitrate setting in the importer which is nice and flexible.

I assume the user is responsible for installing FFmpeg in order for this to work. There is a lot of UX work needed to guide the user to install FFmpeg since we had to do this work in FBX2glTF

i do wonder if this wouldn't be better as an asset lib addon, since then it could come with a LGPL build of FFmpeg for all the platforms (side note: important to build LGPL and not GPL. GPL code is not needed to *decode* h264).
Even if we go forward with including this in the engine, it should be possible to distribute ffmpeg on the asset lib or have a auto download button like we did for FBX2glTF, except the ffmpeg license is much better and safer than the autodesk license so I think we can streamline this

there is something about registering a new extension .theoravideostr which I don't understand... why can't we use whatever .ogv already does, for example read .ogv directly (I assume there would be a ResourceFormatLoader)?

also this PR affects the stop function to call seek(0) which really should be its own PR and bugfix for whatever it is fixing.

berarma commented:
> Just to confirm, because the description doesn't explain what this PR actually does...

Yes, sorry, I will improve the description.

> From what I understand, this adds importer support for mp4/webm/mpg/mpeg/mkv/avi/mov/wmv/qoi extensions? This is pretty innovative and should make videos much easier to use in Godot

Those are the extensions that I thought would be interesting to have, but it could be any format supported by FFmpeg.

> They are transcoded into OGV using some configurable bitrate setting in the importer which is nice and flexible.

Yes, somewhat. Not a bitrate, but a quality setting as used commonly in VBR codecs. It can also scale the video and replace the alpha channel with a color.

> I assume the user is responsible for installing FFmpeg in order for this to work. There is a lot of UX work needed to guide the user to install FFmpeg since we had to do this work in FBX2glTF

Yes, the user should have FFmpeg already installed. Installing FFmpeg is actually very easy, it's a matter of unzipping a file, setting the path in Godot and done.

> i do wonder if this wouldn't be better as an asset lib addon, since then it could come with a LGPL build of FFmpeg for all the platforms (side note: important to build LGPL and not GPL. GPL code is not needed to _decode_ h264). Even if we go forward with including this in the engine, it should be possible to distribute ffmpeg on the asset lib or have a auto download button like we did for FBX2glTF, except the ffmpeg license is much better and safer than the autodesk license so I think we can streamline this

I like the flexibility of being able to install any FFmpeg version. I don't think licenses are a problem since we're not linking any code, but running an external command instead. At the same time, I agree with anything that users find easier to use it.

> there is something about registering a new extension .theoravideostr which I don't understand... why can't we use whatever .ogv already does, for example read .ogv directly (I assume there would be a ResourceFormatLoader)? EDIT: I understand why you want a separate resource type since it precludes needing to load the whole video file in memory. My question is just why does this require a different file extension compared to the existing OGV importer

I think the current OGV importer doesn't use any resource file, it uses the video file directly. Maybe I could do the same, it's the first time I'm looking at the resource system and there are still some things I don't fully understand. I'll take a look at the way OGV files are imported and maybe I can so something similar.

> also this PR affects the stop function to call seek(0) which really should be its own PR and bugfix for whatever it is fixing.

Yes, I agree. For easier testing I prefer to keep it here, and I will create a new PR with only that fix. I just didn't get to do it yet.

Re-importing a file being used in a scene calls `VideoStreamPlayback::stop()` and that would call `seek`. That `seek` call when reloading the resource in the re-import process isn't needed, and freezes the editor. Moving the `seek` to `VideoStreamPlayer::stop()` fixes this issue, and it's in accordance with  the API docs.

fire commented:
> also this PR affects the stop function to call seek(0) which really should be its own PR and bugfix for whatever it is fixing.

We can probably speedly review the stop/seek bugfix if it's in a separate pr. If it's in this pull request, we'll have to wait for this to be complete which will delay the merge.

berarma commented:
I mistakenly had thought FFmpeg would allow qoi frames into some video container but it does not. I wanted to include more lossless codecs, FFmpeg already supports FFv1 inside MKV though.

berarma committed:
Don't use special video resource

berarma committed:
Fix setting case

berarma commented:
> I believe an extra String import option for additional custom import arguments could be valuable.

Yes, this could be an option. I'll add it if there's no objection.

> Unrelated to this, PR in particular, i have noticed in a project, that when i converted a file with my windows ffmpeg with certain arguents, didn't produce the exact same ogv as another guy who was importing it from Linux (mine was glitchy) I am concerned this feature could hide this fact and people would rely conversion will jsut work for everyone because it works in their machine and then when some CICD exports a project it will create a broken vid. Toughts?

This should be very unusual unless the FFmpeg version (or linked libraries) were different. Any software can have this issue. This PR doesn't change that fact.

berarma committed:
Add encoding_speed option

berarma committed:
Fix import with no complex filters

berarma committed:
Add additional arguments setting

berarma commented:
Yay! The [patch that adds a speed level option to FFmpeg](/FFmpeg/FFmpeg/commit/3cd4fa9b895a359cf8c895f601c0016aae8b2c8f) has just been merged. I've added the option to this importer and set the fastest speed (`4`) as default. Imports will be faster but files will be also much bigger. Maybe speed `3` would be a better default since it compresses a lot better while still being half as fast. We can decide after testing.

The new option will be available in the FFmpeg builds since today. When using a build that doesn't have the option yet, it will be ignored. I had thought about adding a warning to the import tab when the FFmpeg version used doesn't support this option, but I have to learn how to add warnings to import tabs.

I've added a setting to write custom FFmpeg options after the input file and before the automatic options. The order of the arguments matters but I think this could be enough.

I've checked the [FBX2glTF download code](https://github.com/godotengine/godot/commit/e93d991329df7489e77ae4c9d3d686d8c1087dec) and it looks like a good solution. The user would be able to download the FFmpeg binary directly from us so we would have control of the version that is used. Advanced users could still change it to a custom binary.

berarma committed:
Add project setting to disable the importer

berarma committed:
Use String.is_empty()

berarma commented:
> > * [ ]  Create asset library addon or automation to install FFmpeg?
> 
> I think we could gain from having built-in installers for external tools in the editor, but it might be worth creating a general implementation that could be reused for all external tools. I'll look into opening a proposal for this.

Yes, that would be much better than reinventing the wheel every time. If we're heading that path then I'll remove the task from the list.

The other open question is if there's something that can be done about the slow imports. The Encoding Speed setting already helps but there will always be slow imports.

I've implemented a project setting to disable the importer since I don't think there's another way.

fire commented:
Can we change ffmpeg to prefer hardware encoding?

Calinou commented:
> Can we change ffmpeg to prefer hardware encoding?

Hardware Theora encoding isn't a thing, and if it was, it would have lower quality than software encoding (this is common to all hardware encoders out there).

fire commented:
Would we accept a av1 decoder in core?

Calinou commented:
> Would we accept an av1 decoder in core?

We don't have support for AV1 playback yet, but on a purely functional level, I would personally welcome it. However, one issue is that AV1 decoders have a large binary size and are complex to build. Some of them are also written in Rust, and we don't want to impose users to install a Rust toolchain to build Godot.

It's probably best implemented as a GDExtension for the time being.

berarma committed:
Add FFmpeg version check

berarma committed:
Reduce FFmpeg command verbosity

fire commented:
I was looking into https://github.com/nvpro-samples/vk_video_samples and trying to see what we can do. Here's a copy of their readme focusing on AV1.

# decoder

1. Extracts (DEMUX via FFMPEG) compressed video from .mp4, .mkv, .mov and others video containers using AV1 compression formats.
2. The HW video decoder processes textures to Vulkan Video Images that can be directly sampled from Vulkan Samplers (Textures).
3. Using ffmpeg converts the YCbCr (YUV) Images to RGB while sampling the decoded images using the [VK_KHR_sampler_ycbcr_conversion](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_KHR_sampler_ycbcr_conversion.html)

# encoder

The upstream sample says this is work in progress.

1. use the Vulkan video encoding extensions to drive the HW-accelerated video encoder
2. this application encodes YCbCr content
3. AV1 compressed video to a file.

fire commented:
Does anyone know the legal issues using hardware encoders and hardware decoders for av1, hevc and mp4?

fire commented:
I posted a proposal about using Windows Media Foundation libraries on Windows. https://github.com/godotengine/godot-proposals/issues/12864 and has a link to a Godot Engine branch.

aaronfranke commented:
@fire @lyuma Why bundle a library or FFmpeg? Why not require the user to supply their own FFmpeg executabe, like this PR already does, and like we are already doing with Blender?
